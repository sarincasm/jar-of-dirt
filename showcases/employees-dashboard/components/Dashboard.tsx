import React, {useEffect, useState} from 'react'
import Head from 'next/head'
import Typography from '@mui/material/Typography'
import Fab from '@mui/material/Fab'

import EmployeeTable from '../components/EmployeeTable'

const SERVER_URL = process.env.NEXT_PUBLIC_SERVER_URL

const Dashboard: React.FC = () => {
	const [employees, setEmployees] = useState([])
	const [isLoading, setLoading] = useState(false)

	useEffect(() => {
		setLoading(true)
		fetch(SERVER_URL + '/employees')
			.then((res) => res.json())
			.then((employees) => {
				setEmployees(employees)
				setLoading(false)
			})
	}, [])

	function handleStateChange(
		employee_id: string,
		new_state: string
	) {
		const data = {new_state}
		fetch(`${SERVER_URL}/employees/${employee_id}`, {
			method: 'PATCH',
			body: JSON.stringify(data),
			headers: {
				'Content-Type': 'application/json',
			},
		})
			.then((res) => res.json())
			.then((data) => {
				setEmployees(data)
				setLoading(false)
			})
	}

	function addNewEmployee() {
		fetch(`${SERVER_URL}/employees`, {
			method: 'POST',
		})
			.then((res) => res.json())
			.then((data) => {
				setEmployees(data)
				setLoading(false)
			})
	}

	if (isLoading) return <p>Loading...</p>
	if (!employees || !employees.length) return <p>No Employee Data</p>

	return (
		<>
			<Head>
				<title>Employee Dashboard</title>
			</Head>
			<Typography align="center" variant="h4" gutterBottom>
				Employee Dashboard
			</Typography>
			<EmployeeTable
				employees={employees}
				onChange={handleStateChange}
			/>
			<Fab
				aria-label="add"
				sx={{
					position: 'fixed',
					bottom: '2%',
					left: '3%',
					backgroundColor: '#0157D9',
				}}
				onClick={addNewEmployee}
			>
				+
			</Fab>
		</>
	)
}

export default Dashboard
