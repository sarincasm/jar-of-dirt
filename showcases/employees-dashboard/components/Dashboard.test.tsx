import React from 'react'
import {
	render,
	act,
	getByText,
	waitFor,
	queryAllByText,
} from '@testing-library/react'

import Dashboard from './Dashboard'

const mockEmployees = [
	{employee_id: '0', current_state: 'ADDED'},
	{employee_id: '1', current_state: 'IN-CHECK'},
	{employee_id: '2', current_state: 'ACTIVE'},
	{employee_id: '3', current_state: 'APPROVED'},
	{employee_id: '4', current_state: 'INACTIVE'},
]

const mockFetch = jest.fn(() =>
	Promise.resolve({
		json: () => Promise.resolve(mockEmployees),
	})
)
// @ts-ignore
global.fetch = mockFetch

describe('Tests for Dashboard', () => {
	afterEach(() => {
		jest.clearAllMocks()
	})
	test('Dashboard renders Loading state', async () => {
		act(() => {
			render(<Dashboard />)
		})
		getByText(document.body, /loading/i)
		expect(mockFetch).toBeCalledTimes(1)
		const mockFetchArgs = mockFetch.mock.calls[0]
		expect(mockFetchArgs).toMatchSnapshot()
		await waitFor(() => getByText(document.body, /dashboard/i))
	})

	test('Dashboard renders initial list of employees', async () => {
		act(() => {
			render(<Dashboard />)
		})
		await waitFor(() => getByText(document.body, /dashboard/i))
		const rowCount = queryAllByText(
			document.body,
			/approved/i
		).length
		expect(rowCount).toBe(5)
	})

	test('Changing state calls the API', async () => {
		act(() => {
			render(<Dashboard />)
		})
		await waitFor(() => getByText(document.body, /dashboard/i))
		const approvedButton = queryAllByText(
			document.body,
			/approved/i
		)[0]
		act(() => {
			approvedButton.click()
		})
		expect(mockFetch).toBeCalledTimes(2)
		const mockFetchArgs = mockFetch.mock.calls[1]
		expect(mockFetchArgs).toMatchSnapshot()
		await waitFor(
			() =>
				new Promise((resolve) => {
					setTimeout(() => {
						resolve('')
					})
				})
		)
	})

	test('Clicking Add button adds an employee', async () => {
		act(() => {
			render(<Dashboard />)
		})
		await waitFor(() => getByText(document.body, /dashboard/i))
		const addButton = getByText(document.body, '+')
		mockFetch.mockResolvedValueOnce({
			json: () =>
				Promise.resolve([
					...mockEmployees,
					{employee_id: '5', current_state: 'ADDED'},
				]),
		})
		act(() => {
			addButton.click()
		})

		expect(mockFetch).toBeCalledTimes(2)
		const mockFetchArgs = mockFetch.mock.calls[1]
		expect(mockFetchArgs).toMatchSnapshot()
		await waitFor(() => getByText(document.body, '5'))
	})
})
