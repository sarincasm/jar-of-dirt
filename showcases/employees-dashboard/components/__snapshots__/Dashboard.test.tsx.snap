// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Dashboard Changing state calls the API 1`] = `
Array [
  "http://localhost:4000/employees/0",
  Object {
    "body": "{\\"new_state\\":\\"APPROVED\\"}",
    "headers": Object {
      "Content-Type": "application/json",
    },
    "method": "PATCH",
  },
]
`;

exports[`Tests for Dashboard Clicking Add button adds an employee 1`] = `
Array [
  "http://localhost:4000/employees",
  Object {
    "method": "POST",
  },
]
`;

exports[`Tests for Dashboard Dashboard renders Loading state 1`] = `
Array [
  "http://localhost:4000/employees",
]
`;
